src/api/swg-client-event-manager-api.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright 2018 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+
+/** @enum {number}  */
+export const FilterResult = {
+  /** the event is allowed to proceed to the listeners */
+  PROCESS_EVENT: 0,
+  /** the event is canceled and the listeners are not informed about it */
+  CANCEL_EVENT: 1,
+};
+
+/**
+ * Defines a client event in SwG
+ * Properties:
+ * - eventType: Required. The AnalyticsEvent type that occurred.
+ * - eventOriginator: Required.  The codebase that initiated the event.
+ * - additionalParameters: Optional.  A JSON object to store generic data.
+ * - isFromUserAction: Optional.  True if the user took an action to generate
+ *   the event.
+ *
+ *  @typedef {{
+ *    eventType: (!AnalyticsEvent),
+ *    eventOriginator: (!EventOriginator),
+ *    isFromUserAction: (?boolean),
+ *    additionalParameters: (?Object),
+ * }}
+ */
+export let SwgClientEvent;
+
+/**
+ * @interface
+ */
+export class SwgClientEventManagerApi {
+  /**
+   * Call this function to log an event. The registered listeners will be
+   * invoked unless the event is filtered.
+   * @param {!function(!SwgClientEvent)} callback
+   */
+  registerEventListener(callback) { }
+
+  /**
+   * Register a filterer for events if you need to potentially prevent the
+   * listeners from hearing about it.  A filterer should return
+   * FilterResult.CANCEL_EVENT to prevent listeners from hearing about the
+   * event.
+   * @param {!function(!SwgClientEvent):FilterResult} callback
```

Comment:
Let's discuss: we could follow are more familiar browser `Event` model where instead of callback result, a client could call a method on the event object itself. E.g. https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault

But in all fairness, I don't think I quite understand the idea behind the `FilterResult`. Why not just do the same check in the `registerEventListener`?

Reply:
The idea is that you have no idea what order each listener is called in.  So if something registered as a listener and canceled the event then its possible that something else already processed that event.  By having an explicit filterer step that comes before the listeners you ensure that there is consistency: either every listener is called or none.  I do not want to seem to provide a method of canceling the event during the listening phase when I am not going to honor that request.  

Which brings me to what I absolutely hated about JavaScript events in general: I could never figure out what was going on with all these different stops and prevents.  I think there were 3 from what I recall (stopPropagation, something similar to that but also somehow different and preventDefault); maybe some of those were provided by jQuery.  And the actual result depended on exactly what kind of event it was and which browser it was and it was all very confusing and resulted in inconsistent difficult to code for behavior that I was never entirely confident would work in every browser and every situation (when I actually had to stop an event from propagating anyway which was very rare).  

preventDefault I am definitely opposed to, stopPropagation would be a bit better but still has the weakness of giving you the seeming ability to halt the event in the middle of the listening phase which would be very bad because the event might have already been logged to a server somewhere before you got a chance to cancel it.

Reply:
Ok. I think I understand this better now. Thanks for the explanation.